### PR TITLE
Simple velocity distortion

### PR DESCRIPTION
**Pull request by PiotrMrozik**

### CODE REVIEW

File: test/src/graph/nodes/VelocityDistortRaysNodeTest.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	rgl_vec3f angularVelocity = { 0.0f, 0.0f, 1.0f };
+
+	EXPECT_RGL_SUCCESS(rgl_node_rays_velocity_distort(&velocityDistortRaysNode, &velocity, &angularVelocity));
+	ASSERT_THAT(velocityDistortRaysNode, testing::NotNull());
+
+	// If (*node) != nullptr
+	EXPECT_RGL_SUCCESS(rgl_node_rays_velocity_distort(&velocityDistortRaysNode, &velocity, &angularVelocity));
+}
+
+TEST_F(VelocityDistortRaysNodeTest, use_case)
```
Comment (msz-rai) [inline test/src/graph/nodes/VelocityDistortRaysNodeTest.cpp:36]: Lack of tests:
- angular velocity
- linear and angular velocities combined
- pipeline with transform node (before / after distortion)
- distance field calculation
It is OK if these tests would be added in another PR.

**Review by msz-rai — CHANGES_REQUESTED** [review]
Good job :+1: 
Please address the comments.
I don't like the PR name also. I propose `Simple velocity distortion`.

File: src/api/apiCore.cpp [hunk 2]
```diff
@@ -659,6 +683,33 @@ void TapePlayer::tape_node_rays_transform(const YAML::Node& yamlNode)
...
+		RGL_API_LOG("rgl_node_rays_velocity_distort(node={}, linearVelocity={}, angularVelocity={})",
+					repr(node), repr(linearVelocity), repr(angularVelocity));
+				CHECK_ARG(node != nullptr);
+				CHECK_ARG(linearVelocity != nullptr);
+				CHECK_ARG(angularVelocity != nullptr);
+
+		createOrUpdateNode<VelocityDistortRaysNode>(node,
+													reinterpret_cast<const Vec3f*>(linearVelocity),
+													reinterpret_cast<const Vec3f*>(angularVelocity));
+	});
```
Comment (PiotrMrozik) [inline src/api/apiCore.cpp:699]:
```suggestion
		createOrUpdateNode<VelocityDistortRaysNode>(node,
												reinterpret_cast<const Vec3f*>(linearVelocity),
												reinterpret_cast<const Vec3f*>(angularVelocity));
	});
```

File: src/graph/VelocityDistortRaysNode.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	auto offsets = input->getTimeOffsets();
+	distortedRays->resize(getRayCount());
+	gpuVelocityDistortRays(stream, getRayCount(),
+						   input->getRays()->getDevicePtr(),
+						   offsets.value()->getDevicePtr(),
+						   offsets.value()->getCount(),
+						   sensorLinearVelocity, sensorAngularVelocity,
+						   distortedRays->getDevicePtr());
+
+}
```
Comment (PiotrMrozik) [inline src/graph/VelocityDistortRaysNode.cpp:44]:
```suggestion
	gpuVelocityDistortRays(stream, getRayCount(),input->getRays()->getDevicePtr(),
					offsets.value()->getDevicePtr(),
					offsets.value()->getCount(),
					sensorLinearVelocity, sensorAngularVelocity,
					distortedRays->getDevicePtr());
}
```

File: include/rgl/api/core.h [hunk 4]
```diff
@@ -481,6 +493,22 @@ rgl_node_rays_set_ring_ids(rgl_node_t* node, const int32_t* ring_ids, int32_t ri
 RGL_API rgl_status_t
 rgl_node_rays_transform(rgl_node_t* node, const rgl_mat3x4f* transform);
 
+/**
+ * Creates or modifies VelocityDistortRaysNode.
+ * The node distorts rays according to the linear velocity and angular velocity of the sensor.
+ * The distortion is performed in the sensor-local coordinate frame.
+ * This node requires that rays time offsets are set.
+ * The distortion takes int account only sensor velocity.
+ * The velocity of the objects being scanned by the sensor is not taken into consideration.
```
Comment (PiotrMrozik) [inline include/rgl/api/core.h:502]:
```suggestion
 * Creates or modifies VelocityDistortRaysNode.
 * The node distorts rays according to the linear velocity and angular velocity of the sensor.
 * NOTE! 
 * The distortion is performed in the sensor-local coordinate frame. Concidering that, This node has to be placed before TransformRaysNode. If not velocity and angular velocity have to be provided in global coordinates frame.
 * This node requires that rays time offsets are set.
 * The distortion takes int account only sensor velocity.
 * The velocity of the objects being scanned by the sensor is not taken into consideration.
```

File: include/rgl/api/core.h [hunk 5]
```diff
@@ -481,6 +493,25 @@ rgl_node_rays_set_ring_ids(rgl_node_t* node, const int32_t* ring_ids, int32_t ri
 RGL_API rgl_status_t
 rgl_node_rays_transform(rgl_node_t* node, const rgl_mat3x4f* transform);
 
+/**
+ * Creates or modifies VelocityDistortRaysNode.
+ * The node distorts rays according to the sensor's linear velocity and angular velocity.
+ * NOTE! 
+ * The distortion is performed in the sensor-local coordinate frame. 
+ * Considering that, This node has to be placed before TransformRaysNode. 
+ * If not velocity and angular velocity must be provided in the global coordinates frame.
```
Comment (msz-rai) [inline include/rgl/api/core.h:502]: I don't like this description.

> in the global coordinates frame.

We cannot assume that TransformRaysNode is always transforming to the global coordinate frame.

I'd describe it as:
`The velocities passed to that node must match the coordinate frame in which rays are described/transformed. E.g. ... (here put some examples)`

File: include/rgl/api/core.h [hunk 6]
```diff
@@ -481,6 +493,25 @@ rgl_node_rays_set_ring_ids(rgl_node_t* node, const int32_t* ring_ids, int32_t ri
...
+ * This node requires that rays time offsets are set.
+ * The distortion takes into account only sensor velocity.
+ * The velocity of the objects being scanned by the sensor is not considered.
+ * Graph input: rays
+ * Graph output: rays
+ * @param node If (*node) == nullptr, a new node will be created. Otherwise, (*node) will be modified.
+ * @param velocity Pointer to a single 3D vector describing the velocity to perform distortion. The velocity is in meters per second.
+ * @param angularVelocity Pointer to a single 3D vector describing the delta angular velocity in radians to perform distortion in euler angles (roll, pitch, yaw).
```
Comment (msz-rai) [inline include/rgl/api/core.h:510]: radians is not an angular velocity unit.

File: src/gpu/nodeKernels.hpp [hunk 7]
```diff
@@ -33,6 +33,7 @@ void gpuFindCompaction(cudaStream_t, size_t pointCount, const Field<IS_HIT_I32>:
 void gpuFormatSoaToAos(cudaStream_t, size_t pointCount, size_t pointSize, size_t fieldCount, const GPUFieldDesc *soaInData, char *aosOutData);
 void gpuFormatAosToSoa(cudaStream_t, size_t pointCount, size_t pointSize, size_t fieldCount, const char* aosInData, GPUFieldDesc* soaOutData);
 void gpuTransformRays(cudaStream_t, size_t rayCount, const Mat3x4f* inRays, Mat3x4f* outRays, Mat3x4f transform);
+void gpuVelocityDistortRays(cudaStream_t stream, size_t rayCount, const Mat3x4f* inRays, const float* offsets, size_t offsetsCount, const Vec3f velocity, const Vec3f angularVelocity, Mat3x4f* outRays);
```
Comment (msz-rai) [inline src/gpu/nodeKernels.hpp:36]:
```suggestion
void gpuVelocityDistortRays(cudaStream_t stream, size_t rayCount, const Mat3x4f* inRays, const float* offsets, size_t offsetsCount, const Vec3f sensorLinearVelocity, const Vec3f sensorAngularVelocity, Mat3x4f* outRays);
```

File: src/graph/NodesCore.hpp [hunk 8]
```diff
@@ -183,6 +183,24 @@ struct TransformRaysNode : Node, IRaysNodeSingleInput
 	VArrayProxy<Mat3x4f>::Ptr rays = VArrayProxy<Mat3x4f>::create();
 };
 
+struct VelocityDistortRaysNode : Node, IRaysNodeSingleInput
+{
+	using Ptr = std::shared_ptr<VelocityDistortRaysNode>;
+	void setParameters(const Vec3f* velocity, const Vec3f* angularVelocity);
```
Comment (msz-rai) [inline src/graph/NodesCore.hpp:189]:
```suggestion
	void setParameters(const Vec3f* linearVelocity, const Vec3f* angularVelocity);
```

File: src/graph/NodesCore.hpp [hunk 9]
```diff
@@ -223,6 +243,25 @@ struct SetRingIdsRaysNode : Node, IRaysNodeSingleInput
 	VArrayProxy<int>::Ptr ringIds = VArrayProxy<int>::create();
 };
 
+struct SetTimeOffsetsRaysNode : Node, IRaysNodeSingleInput
+{
+	using Ptr = std::shared_ptr<SetTimeOffsetsRaysNode>;
+	void setParameters(const float* raysTimeOffests, size_t timeOffsetsCount);
```
Comment (msz-rai) [inline src/graph/NodesCore.hpp:249]:
```suggestion
	void setParameters(const float* raysTimeOffsets, size_t timeOffsetsCount);
```

File: src/graph/SetTimeOffsetsRaysNode.cpp [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+{
+	timeOffsets->setData(raysTimeOffsetsRaw, raysTimeOffsetsCount);
+}
+
+void SetTimeOffsetsRaysNode::validate()
+{
+	input = getValidInput<IRaysNode>();
+
+	if (input->getRayCount() != timeOffsets->getCount()) {
+		auto msg = fmt::format("offsets don't match number of rays. "
```
Comment (msz-rai) [inline src/graph/SetTimeOffsetsRaysNode.cpp:27]:
```suggestion
		auto msg = fmt::format("time offsets don't match number of rays. "
```

File: test/src/graph/nodes/VelocityDistortRaysNodeTest.cpp [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+	// Run graph again with velocity distortion
+	EXPECT_RGL_SUCCESS(rgl_graph_run(raytrace));
+
+	EXPECT_RGL_SUCCESS(rgl_graph_get_result_data(raytrace, XYZ_F32, outPoints.data()));
+
+	// After distortion rays should hit with 0.1 offset in y axis, without changes in x and z axis.(
+	for (int i = 0; i < rayCount; ++i)
+	{
+		EXPECT_NEAR(outPoints[i].x(), outPoints.begin()->x(), EPSILON_F);
+		EXPECT_NEAR(outPoints[i].y(), outPoints.begin()->y() + ( i * rayTimeOffset), EPSILON_F);
```
Comment (msz-rai) [inline test/src/graph/nodes/VelocityDistortRaysNodeTest.cpp:106]: Why sensor velocity is not considered in this equation?

File: test/src/graphTest.cpp [hunk 12]
```diff
@@ -17,13 +17,17 @@ TEST_F(GraphCase, FullLinear)
...
 
 	std::vector<rgl_mat3x4f> rays = makeLidar3dRays(360, 180, 0.36, 0.18);
+	std::vector<float> timeOffsets = std::vector<float>(rays.size(), 0.0f);
 	rgl_mat3x4f lidarPoseTf = Mat3x4f::TRS({5, 5, 5}, {45, 45, 45}).toRGL();
 	rgl_mat3x4f shearTf = Mat3x4f::shear({0,0}, {-1, -1}, {0, 0}).toRGL();
+	rgl_vec3f testVelocity = {1, 2, 3};
 
 	EXPECT_RGL_SUCCESS(rgl_node_rays_from_mat3x4f(&useRays, rays.data(), rays.size()));
+	EXPECT_RGL_SUCCESS(rgl_node_rays_set_time_offsets(&setOffsets, timeOffsets.data(), timeOffsets.size()));
+	EXPECT_RGL_SUCCESS(rgl_node_rays_velocity_distort(&velocityDistort, &testVelocity, &testVelocity));
```
Comment (msz-rai) [inline test/src/graphTest.cpp:30]: I don't understand this change

**Review by msz-rai — CHANGES_REQUESTED** [review]
I have left some suggestions.
Please validate units when distorting rays.

File: include/rgl/api/core.h [hunk 13]
```diff
@@ -481,6 +493,25 @@ rgl_node_rays_set_ring_ids(rgl_node_t* node, const int32_t* ring_ids, int32_t ri
...
+ * If node is before TransformRaysNode, then the velocities must be in the sensor-local coordinate frame.
+ * Otherwise, user hase to transform the velocities to TransformRaysNode coordinate frame.
+ * This node requires that rays time offsets are set.
+ * The distortion takes into account only sensor velocity.
+ * The velocity of the objects being scanned by the sensor is not considered.
+ * Graph input: rays
+ * Graph output: rays
+ * @param node If (*node) == nullptr, a new node will be created. Otherwise, (*node) will be modified.
+ * @param velocity velocity Pointer to a single 3D vector describing the linear velocity of the sensor. The velocity is in meters per second.
+ * @param angularVelocity Pointer to a single 3D vector describing the delta angular velocity in radians per secondto perform distortion in euler angles (roll, pitch, yaw).
```
Comment (msz-rai) [inline include/rgl/api/core.h:510]:
```suggestion
 * @param angularVelocity Pointer to a single 3D vector describing the angular velocity of the sensor in euler angles (roll, pitch, yaw). The velocity is in radians per second.
```

File: src/api/apiCore.cpp [hunk 14]
```diff
@@ -659,6 +683,35 @@ void TapePlayer::tape_node_rays_transform(const YAML::Node& yamlNode)
 	rgl_node_rays_transform(&node, reinterpret_cast<const rgl_mat3x4f*>(fileMmap + yamlNode[1].as<size_t>()));
 	tapeNodes.insert({nodeId, node});
 }
+RGL_API rgl_status_t
+rgl_node_rays_velocity_distort(rgl_node_t* node, const rgl_vec3f* linear_velocity, const rgl_vec3f* angular_velocity)
+{
+	auto status = rglSafeCall([&]() {
+		RGL_API_LOG("rgl_node_rays_velocity_distort(node={}, linearVelocity={}, angularVelocity={})",
```
Comment (msz-rai) [inline src/api/apiCore.cpp:690]:
```suggestion
		RGL_API_LOG("rgl_node_rays_velocity_distort(node={}, linear_velocity={}, angular_velocity={})",
```

File: src/gpu/nodeKernels.cu [hunk 15]
```diff
@@ -43,6 +43,12 @@ __global__ void kTransformRays(size_t rayCount, const Mat3x4f* inRays, Mat3x4f*
 	outRays[tid] = transform * inRays[tid];
 }
 
+__global__ void kVelocityDistortRays(size_t rayCount, const Mat3x4f* inRays, const float* offsets, size_t offsetsCount, const Vec3f sensorLinearVelocity, const Vec3f sensorAngularVelocity, Mat3x4f* outRays)
+{
+	LIMIT(rayCount);
+	outRays[tid] = Mat3x4f::TRS((offsets[tid] * (sensorLinearVelocity * 0.001f)) * 1000.0f, (offsets[tid] * (sensorAngularVelocity) * 0.001f) * 1000.0f, Vec3f(1.0f, 1.0f, 1.0f)) * inRays[tid];
```
Comment (msz-rai) [inline src/gpu/nodeKernels.cu:49]: Please put a comment that describes why we are performing unit conversions in this way.

File: src/gpu/nodeKernels.hpp [hunk 16]
```diff
@@ -33,6 +33,7 @@ void gpuFindCompaction(cudaStream_t, size_t pointCount, const Field<IS_HIT_I32>:
 void gpuFormatSoaToAos(cudaStream_t, size_t pointCount, size_t pointSize, size_t fieldCount, const GPUFieldDesc *soaInData, char *aosOutData);
 void gpuFormatAosToSoa(cudaStream_t, size_t pointCount, size_t pointSize, size_t fieldCount, const char* aosInData, GPUFieldDesc* soaOutData);
 void gpuTransformRays(cudaStream_t, size_t rayCount, const Mat3x4f* inRays, Mat3x4f* outRays, Mat3x4f transform);
+void gpuVelocityDistortRays(cudaStream_t stream, size_t rayCount, const Mat3x4f* inRays, const float* offsets, size_t offsetsCount, const Vec3f sensorVelocity, const Vec3f sensorAngularVelocity, Mat3x4f* outRays);
```
Comment (msz-rai) [inline src/gpu/nodeKernels.hpp:36]:
```suggestion
void gpuVelocityDistortRays(cudaStream_t stream, size_t rayCount, const Mat3x4f* inRays, const float* offsets, size_t offsetsCount, const Vec3f sensorLinearVelocity, const Vec3f sensorAngularVelocity, Mat3x4f* outRays);
```

File: test/src/apiSurfaceTests.cpp [hunk 17]
```diff
@@ -110,3 +110,16 @@ TEST_F(APISurfaceTests, rgl_entity_set_pose)
...
+	rgl_entity_t entity = makeEntity();
+	int validID = 1;
+
+	// Invalid args
+	EXPECT_RGL_INVALID_ARGUMENT(rgl_entity_set_id(nullptr, 1), "entity != nullptr");
+	EXPECT_RGL_INVALID_ARGUMENT(rgl_entity_set_id(entity, RGL_ENTITY_INVALID_ID), "id != RGL_ENTITY_INVALID_ID");
+
+	// Correct set_pose
+	EXPECT_RGL_SUCCESS(rgl_entity_set_id(entity, validID));
+}
```
Comment (msz-rai) [inline test/src/apiSurfaceTests.cpp:125]: Not related to velocity distortion

**Review by msz-rai — APPROVED** [review]
LGTM

**Pull request merged**